File: Task 4/Java/app/src/main/java/dev/ogabek/java/manager/StorageManager.java
package dev.ogabek.java.manager;

import android.content.Context;
import android.net.Uri;

import androidx.annotation.NonNull;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;

public class StorageManager {

    static FirebaseStorage storage = FirebaseStorage.getInstance();
    static StorageReference storageRef = storage.getReference();
    static StorageReference photoRef = storageRef.child("photos");

    public static void uploadPhotos(Context context, Uri uri, StorageHandler handler){
        String type = context.getContentResolver().getType(uri);
        type = type.substring(type.lastIndexOf("/") + 1);
        String fileName = System.currentTimeMillis() + "." + type;
        UploadTask uploadTask = photoRef.child(fileName).putFile(uri);
        uploadTask.addOnSuccessListener(taskSnapshot -> {
            Task<Uri> result = taskSnapshot.getMetadata().getReference().getDownloadUrl();
            result.addOnSuccessListener( it -> {
                String imgUrl = it.toString();
                handler.onSuccess(imgUrl);
            }).addOnFailureListener(handler::onError);
        }).addOnFailureListener(handler::onError);
    }

}

File: Task 2/Kotlin/app/src/main/java/dev/ogabek/kotlin/manager/StorageManager.kt
package dev.ogabek.kotlin.manager

import android.content.Context
import android.net.Uri
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.UploadTask

class StorageManager {

    companion object{

        val storage = FirebaseStorage.getInstance()
        val storageRef = storage.reference
        val photoRef = storageRef.child("photos")

        fun uploadPhotos(context: Context, uri: Uri, handler: StorageHandler){
            var type: String? = context.contentResolver.getType(uri)
            type = type!!.substring(type.lastIndexOf("/") + 1)
            val fileName = System.currentTimeMillis().toString() + ".${type}"
            val uploadTask: UploadTask = photoRef.child(fileName).putFile(uri)
            uploadTask.addOnSuccessListener { it ->
                val result = it.metadata!!.reference!!.downloadUrl
                result.addOnSuccessListener {
                    val imgUrl = it.toString()
                    handler.onSuccess(imgUrl)
                }.addOnFailureListener { e ->
                    handler.onError(e)
                }
            }.addOnFailureListener{ e ->
                handler.onError(e)
            }
        }
    }
}